pack/storage.go

```diff
@@ -0,0 +1,70 @@
+package pack
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+)
+
+type Storage struct {
+	packs *Set
+}
+
+func NewStorage(root string) (*Storage, error) {
+	packs, err := NewSet(root)
+	if err != nil {
+		return nil, err
+	}
+	return &Storage{packs: packs}, nil
+}
+
+func (f *Storage) Open(oid []byte) (r io.ReadWriteCloser, err error) {
+	obj, err := f.packs.Object(oid)
+	if err != nil {
+		return nil, err
+	}
+	return &delayedObjectReadWriter{obj: obj}, nil
+}
+
+func (f *Storage) Close() error {
+	return f.packs.Close()
+}
+
+// IsCompressed returns false, because data returned is already decompressed.
+func (f *Storage) IsCompressed() bool {
+	return false
+}
+
+type delayedObjectReadWriter struct {
+	obj *Object
+	mr  io.Reader
+}
+
+func (d *delayedObjectReadWriter) Read(b []byte) (int, error) {
+	if d.mr == nil {
+		data, err := d.obj.Unpack()
+		if err != nil {
+			return 0, err
+		}
+		d.mr = io.MultiReader(
+			// Git object header:
+			strings.NewReader(fmt.Sprintf("%s %d\x00",
+				d.obj.Type(), len(data),
+			)),
+
+			// Git object (uncompressed) contents:
+			bytes.NewReader(data),
+		)
+	}
+	return d.mr.Read(b)
+}
+
+func (d *delayedObjectReadWriter) Write(b []byte) (int, error) {
```

Comment:
This is because `Open` returns an `io.ReadWriteCloser` and we're reusing the `Open` method from the `storer` interface.  I'm not completely clear on why we need an `io.ReadWriteCloser` instead of an `io.ReadCloser`, but let me spend some more time looking at it to see if I can adjust it to use `io.ReadCloser` instead.